### DISCUSSION TITLE
How to create request body logging middleware?

### DISCUSSION THREAD

**Discussion by alloc33** [rwf2/Rocket#2496] [Questions]
I tried to find it in docs but nothing helpful was there regarding this particular case. I want to create I simple middleware (fairing) which would print request body parameters if request body exists. How can I do it?

**Comment by SergioBenitez** [comment]
> request body parameters

What do you mean by this?

If you want to log the entire request body, that isn't currently possible as Rocket enforces that the request body ([`Data`](https://api.rocket.rs/v0.5-rc/rocket/struct.Data.html)) is streamed. Logging the body would necessitate 1) buffering the entire, possibly infinite request body in memory, a trivial vector for a DoS attack, or 2) preventing handlers down the line from reading the request body.

If you're okay with only logging _some_ part of the request body, you can use a fairing along with [`Data::peek()`](https://api.rocket.rs/v0.5-rc/rocket/data/struct.Data.html#method.peek):

```rust
rocket::build().attach(AdHoc::on_request("Body Logger", |_req, data| Box::pin(async {
   log!("{:?}", data.peek(512).await);
})));
```
